File: scripts/Linked List Cycle II/Linked List Cycle II.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) return null;
        ListNode tortoise = head;
        ListNode hare = new ListNode();
        hare.next = head.next;
        while (hare != null && hare.next != null && hare != tortoise) {
            tortoise = tortoise.next;
            hare = hare.next.next;
        }
        if (hare == null || hare.next == null) return null;
        tortoise = head;
        while (tortoise != hare) {
            tortoise = tortoise.next;
            hare = hare.next;
        }
        return hare;
    }
}

File: scripts/algorithms/L/Linked List Cycle II/Linked List Cycle II.js
// Runtime: 94 ms (Top 78.70%) | Memory: 44.9 MB (Top 73.58%)
var detectCycle = function(head) {
    let slowRunner = head
    let fastRunner = head

    while(fastRunner) {
        fastRunner = fastRunner.next?.next
        slowRunner = slowRunner.next

        if (fastRunner === slowRunner) {
            fastRunner = slowRunner
            slowRunner = head

            while (slowRunner !== fastRunner) {
                slowRunner = slowRunner.next
                fastRunner = fastRunner.next
            }

            return slowRunner
        }
    }

    return null
};